common/main/java/com/couchbase/lite/ResultSet.java

```diff
@@ -148,11 +145,30 @@ public List<Result> allResults() {
     public void close() {
         final C4QueryEnumerator qEnum;
         synchronized (lock) {
-            if (c4enum == null) { return; }
             qEnum = c4enum;
             c4enum = null;
         }
-        synchronized (getDbLock()) { qEnum.close(); }
+        if (qEnum == null) { return; }
+
+        final AbstractDatabase db = context.getDatabase();
+        if (db == null) { throw new IllegalStateException("Could not obtain db lock"); }
+
+        synchronized (db.getDbLock()) { qEnum.close(); }
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        try {
+            final C4QueryEnumerator qEnum = c4enum;
+
+            // !!! This absolutely should hold the db lock the way close() does.
+            // If it does, though, it stands a good chance of timing out the finalizer thread.
+            // Cross your fingers and pray.
```

Comment:
This comment scares me. Sound like a time bomb for a CBSE that is difficult to know what is going on. 

@borrrden Have you mentioed before that .NET has never locked this and never seen any issues?

Reply:
I think we have to do this.  If we try to seize the lock, we are guaranteed that the finalizer thread will time out.